bundles/pcmjava/tools.vitruv.applications.pcmjava.common/src/tools/vitruv/applications/pcmjava/common/pcm2java/Pcm2JavaCommon.reactions

```diff
@@ -507,6 +508,33 @@ routine changeInnerDeclarationType(pcm::InnerDeclaration innerDeclaration, java:
 // ######################################################
 // ################ JAVA PACKAGE ROUTINES ################
 
+routine createOrFindJavaPackage(EObject sourceElementMappedToPackage, java::Package parentPackage, String packageName, String newTag) {
+	match {
+		require absence of java::Package corresponding to sourceElementMappedToPackage tagged with newTag
+		val allPackages = retrieve many java::Package corresponding to ContainersPackage.Literals.PACKAGE
+	}
+	action {
+		call {
+			var matchingPackages = allPackages.filter[name == packageName.toFirstLower]
+			if(parentPackage !== null) { // checks full package namespace for non-root packages:
+				matchingPackages = matchingPackages.filter[namespacesAsString == parentPackage.namespacesAsString + parentPackage.name + "."]
+			}
```

Comment:
It should be possible to put this into a `with` block of the retrieval. The `with` block is able to define whether a candidate element shall be selected or not, just like the called filter methods. In consequence, there would be a clearer separation between matching and actions.